### PR TITLE
chore(deps): allow psr/http-message v2

### PR DESCRIPTION
**Pull request by simPod**

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
In the CI:

>   - Locking psr/http-message (1.1)

So that's not enough.

Also, is this lib really compatible with v2? Doesn't it miss some return types?

**Comment by simPod** [comment]
Hm, let's test it.

**Comment by simPod** [comment]
There are indeed some type changes required but I believe the v2 support won't be possible now since old php versions are allowed.

File: src/MessageTrait.php [hunk 1]
```diff
@@ -81,17 +77,17 @@ public function getHeaderLine($header): string
         return \implode(', ', $this->getHeader($header));
     }
 
-    public function withHeader($header, $value): self
+    public function withHeader(string $name, $value): self
```
Comment (nicolas-grekas) [inline src/MessageTrait.php:80]: the name should not be changed

File: src/Response.php [hunk 2]
```diff
@@ -65,7 +65,7 @@ public function getReasonPhrase(): string
         return $this->reasonPhrase;
     }
 
-    public function withStatus($code, $reasonPhrase = ''): self
+    public function withStatus(int $code, string $reasonPhrase = ''): ResponseInterface
```
Comment (nicolas-grekas) [inline src/Response.php:68]: `self` should be kept everywhere

File: src/Response.php [hunk 3]
```diff
@@ -65,7 +65,7 @@ public function getReasonPhrase(): string
...
     {
         if (!\is_int($code) && !\is_string($code)) {
```
Comment (nicolas-grekas) [inline src/Response.php:70]: can be removed

File: src/ServerRequest.php [hunk 4]
```diff
@@ -79,7 +79,7 @@ public function getUploadedFiles(): array
     /**
      * @return static
      */
-    public function withUploadedFiles(array $uploadedFiles)
+    public function withUploadedFiles(array $uploadedFiles): ServerRequestInterface
```
Comment (nicolas-grekas) [inline src/ServerRequest.php:82]: we should use `self` everywhere instead

File: src/ServerRequest.php [hunk 5]
```diff
@@ -150,7 +150,7 @@ public function getAttributes(): array
     /**
      * @return mixed
      */
-    public function getAttribute($attribute, $default = null)
+    public function getAttribute(string $attribute, $default = null)
     {
         if (!\is_string($attribute)) {
```
Comment (nicolas-grekas) [inline src/ServerRequest.php:155]: drop

File: src/ServerRequest.php [hunk 6]
```diff
@@ -163,7 +163,7 @@ public function getAttribute($attribute, $default = null)
         return $this->attributes[$attribute];
     }
 
-    public function withAttribute($attribute, $value): self
+    public function withAttribute(string $attribute, $value): ServerRequestInterface
     {
         if (!\is_string($attribute)) {
```
Comment (nicolas-grekas) [inline src/ServerRequest.php:168]: drop

File: src/ServerRequest.php [hunk 7]
```diff
@@ -175,7 +175,7 @@ public function withAttribute($attribute, $value): self
         return $new;
     }
 
-    public function withoutAttribute($attribute): self
+    public function withoutAttribute(string $attribute): ServerRequestInterface
     {
         if (!\is_string($attribute)) {
```
Comment (nicolas-grekas) [inline src/ServerRequest.php:180]: drop

File: src/Stream.php [hunk 8]
```diff
@@ -293,7 +290,7 @@ public function getContents(): string
     /**
      * @return mixed
      */
-    public function getMetadata($key = null)
+    public function getMetadata(string $key = null)
     {
         if (null !== $key && !\is_string($key)) {
```
Comment (nicolas-grekas) [inline src/Stream.php:295]: drop

File: src/Uri.php [hunk 9]
```diff
@@ -138,7 +138,7 @@ public function getFragment(): string
         return $this->fragment;
     }
 
-    public function withScheme($scheme): self
+    public function withScheme(string $scheme): UriInterface
     {
         if (!\is_string($scheme)) {
```
Comment (nicolas-grekas) [inline src/Uri.php:143]: drop (I'm stopping here, please continue :)

File: src/Uri.php [hunk 10]
```diff
@@ -155,7 +155,7 @@ public function withScheme($scheme): self
         return $new;
     }
 
-    public function withUserInfo($user, $password = null): self
+    public function withUserInfo(string $user, ?string $password = null): UriInterface
```
Comment (nicolas-grekas) [inline src/Uri.php:158]: self everywhere, please continue :)

**Comment by nicolas-grekas** [comment]
We should bump to PHP >=7.2 also

**Comment by simPod** [comment]
If that's okay with you I've opened `>=7.2 PR` here https://github.com/Nyholm/psr7/pull/226

**Comment by simPod** [comment]
Also, I think we cannot have support for both v1 and v2 since methods with typed arguments are not backward compatible.

**Comment by nicolas-grekas** [comment]
> methods with typed arguments are not backward compatible

because of the behavior change when passing incompatible types, or?

**Comment by simPod** [comment]
Because changing e.g.

```diff
-     public function withProtocolVersion($version): self
+     public function withProtocolVersion(string $version): self
```

is required and compatible with v2 but not compatible with v1

**Comment by nicolas-grekas** [comment]
It's compatible with 1.1, so we could just bump to ^1.1|^2.0, no?

File: src/MessageTrait.php (hunk 1 above)
Comment (simPod) [reply src/MessageTrait.php:80]: Ok but it has been changed in psr/http-message v2

Comment (nicolas-grekas) [reply src/MessageTrait.php:80]: argument names are not part of the contract in PHP

**Comment by simPod** [comment]
Test suite is being adapted here https://github.com/php-http/psr7-integration-tests/pull/68

**Comment by nicolas-grekas** [comment]
Closing in favor of #234, thanks for pushing for this.

**Pull request closed without merging**